lib/krb5/principal.c

```diff
@@ -1260,17 +1260,43 @@ krb5_principal_is_anonymous(krb5_context context,
 {
     int anon_realm;
 
-    if ((p->name.name_type != KRB5_NT_WELLKNOWN &&
-         p->name.name_type != KRB5_NT_UNKNOWN) ||
-        p->name.name_string.len != 2 ||
+    /*
+     * Heimdal versions 7.5 and below left the name-type at KRB5_NT_PRINCIPAL
+     * even with anonymous pkinit responses.  To retain interoperability with
+     * legacy KDCs, the name-type is not checked by the client after requesting
+     * a fully anonymous ticket.
+     */
+    if ((flags & KRB5_ANON_MATCH_NAME_TYPE) &&
+        p->name.name_type != KRB5_NT_WELLKNOWN &&
```

Comment:
If `krb5_principal_is_anonymous()` has shipped (I think it has?) then it would be better not to change the behaviour. Rather, I would suggest inverting the logic so the additional flag is `KRB5_ANON_MATCH_IGNORE_NT` for example.

Reply:
OK, done in next update.